### ISSUE TITLE
Integration between Cube and Fabric8-Maven-Plugin

### ISSUE THREAD

**Issue by lordofthejars** [arquillian/arquillian-organization#34]
Currently, Boosters relies on Fabric8 Maven plugin to generate Kubernetes/OpenShift resources and deploy them into a cluster.

This means that any test that involves deploying the service before running the test, cannot run from IDE since it needs to generate resources and deploy them and this is done in Maven run so not from IDE.

So we need to think about an integration which makes running tests even when F8 plugin is involved.

Probably there are a lot of ways of fixing this, what it comes to my mind as the first approach uses the F8 logic but inside Arquillian lifecycle instead of Maven lifecycle.

Another approach might be let f-m-p generate the resources (so user is forced to run at least once `fabric8:resources`) goal, but then make Cube do the deployment logic.

In any stategy, is important to notice that we need to deal that a test can run from IDE and from Maven (within f-m-p lifecycle).

Cube issue: https://github.com/arquillian/arquillian-cube/issues/806

**Comment by dipak-pawar** [comment]
Something to look for integration of cube with fabric8 maven plugin - https://github.com/fabric8io/fabric8-build

**Comment by lordofthejars** [comment]
@dipak-pawar yes but notice that it seems a bit outdated. 9 months without any commit. Maybe it is normal, but something to figure out before.

**Comment by dipak-pawar** [comment]
I think @rhuss started this, maybe we can discuss with him

**Comment by rhuss** [comment]
@lordofthejars actually, this was the vision at that time, but then I was disconnected from the project because of some changes, so in short: I still believe the vision is a good one, but I cant really comment whether the fabric8-maven-plugin will it eventually pick up.

All I can say is, that for the docker-maven-plugin I still plan to integrate the 'generator' support to allow zero-configuration when building. Not sure whether it follow the outlined plan, though.

**Comment by bartoszmajsak** [comment]
Related issue in Cube https://github.com/arquillian/arquillian-cube/issues/806

**Comment by dipak-pawar** [comment]
After looking into fabric8-maven-plugin code and trying to add some abstraction in fmp such a that we can use `build`, `resource` goal with using abstracted classes. My findings are as follows:
* It depends on some classes from [`docker-maven-plugin`](https://github.com/fabric8io/docker-maven-plugin) which is coupled with maven(MavenSession, pluginManager, MavenProject and many more)
* Both the plugins are using resource injection using `@Component`. We need to take care of it on cube side for initialization. There are few resources injected using `@Component` in [`docker-maven-plugin`](https://github.com/fabric8io/docker-maven-plugin/blob/master/src/main/java/io/fabric8/maven/docker/AbstractBuildSupportMojo.java#L31) and [`fabric8-maven-plugin`](https://github.com/fabric8io/fabric8-maven-plugin/blob/master/plugin/src/main/java/io/fabric8/maven/plugin/mojo/build/ResourceMojo.java#L103)
* Even if we decided to abstract only two goals, we have to abstract everything i.e. enricher, generators, resource configuration, image configuration to get end to end experience using cube.
* Assuming we have abstraction available. We need ProjectContext(baseDirectory, buildDirectory, groupId, artifactId, version,  ....) on cube side to make abstracted code working. 

After all above findings, It needs lot of efforts and refactoring in both the project i.e. fabric8-maven-plugin and docker-maven-plugin.

I would go with embedded maven on cube side for now.

**Comment by dipak-pawar** [comment]
PR with fmp integration for cube - https://github.com/arquillian/arquillian-cube/pull/1013
